src/repo/remote.rs

```diff
@@ -0,0 +1,692 @@
+                )
+            })?;
+
+            let pack_checksum = parts.next().ok_or_else(|| {
+                Error::BadRemoteIndexFormat(
+                    format!("Expected pack checksum, reached end of line {}", line_no),
+                    None,
+                )
+            })?;
+
+            // Pack URL (possibly relative)
+            let relative_url = parts.next().ok_or_else(|| {
+                Error::BadRemoteIndexFormat(
+                    format!("Expected pack URL, reached end of line {}", line_no),
+                    None,
+                )
+            })?;
+
+            // Always provide the index URL as base URL. If the pack URL is
+            // absolute, the `join` will use it as-is.
+            let absolute_url = match base_url.join(relative_url) {
+                Ok(url) => url,
+                Err(_) => {
+                    return Err(Error::BadRemoteIndexFormat(
+                        format!(
+                            "Expected a valid pack URL, found {} on line {}",
+                            url, line_no
+                        ),
+                        None,
+                    ));
+                }
+            };
+
+            if parts.next().is_some() {
+                return Err(Error::BadRemoteIndexFormat(
+                    format!("Too many values on line {}", line_no),
+                    None,
+                ));
+            }
+
+            let index_checksum = hex::decode(index_checksum)
+                .map_err(|_| {
+                    Error::BadRemoteIndexFormat(
+                        format!("Bad pack index checksum format on line {}", line_no),
+                        None,
+                    )
+                })?
+                .try_into()
+                .map_err(|_| {
+                    Error::BadRemoteIndexFormat(
+                        format!(
+                            "The value for pack index checksum on line {} is not the right length",
+                            line_no
+                        ),
+                        None,
+                    )
+                })?;
+
+            let pack_checksum = hex::decode(pack_checksum)
+                .map_err(|_| {
+                    Error::BadRemoteIndexFormat(
+                        format!("Bad pack checksum format on line {}", line_no),
+                        None,
+                    )
+                })?
+                .try_into()
+                .map_err(|_| {
+                    Error::BadRemoteIndexFormat(
+                        format!(
+                            "The value for pack checksum on line {} is not the right length",
+                            line_no
+                        ),
+                        None,
+                    )
+                })?;
+
+            packs.push(RemotePack {
+                url: absolute_url.as_str().to_owned(),
+                index_checksum,
+                pack_checksum,
+            });
+        }
```

Comment:
At least the code as-is is pretty obvious. So don't take action on my suggestion to the detriment of the obviousness.

Nit. The one complaint would be that the line-count explosion due to the format strings isn't great, if you can think of a quick way to shrink those all down to one line.

Reply:
`cargo fmt` does this. I will see if I can figure something out to reduce the level of noise.